executor/write.go

```diff
@@ -788,28 +790,33 @@ type InsertExec struct {
 	Priority  mysql.PriorityEnum
 	IgnoreErr bool
 
-	finished bool
+	finished      bool
+	rowCount      int
+	dirtyRowValue map[string][]byte
 }
 
 func (e *InsertExec) exec(ctx context.Context, rows [][]types.Datum) (Row, error) {
 	// If tidb_batch_insert is ON and not in a transaction, we could use BatchInsert mode.
 	sessVars := e.ctx.GetSessionVars()
 	defer sessVars.CleanBuffers()
-	batchInsert := sessVars.BatchInsert && !sessVars.InTxn()
-	batchSize := sessVars.DMLBatchSize
 
-	txn := e.ctx.Txn()
-	rowCount := 0
-	writeBufs := sessVars.GetWriteStmtBufs()
+	e.rowCount = 0
 	if !sessVars.ImportingData {
-		writeBufs.BufStore = kv.NewBufferStore(txn, kv.TempTxnMemBufCap)
+		sessVars.GetWriteStmtBufs().BufStore = kv.NewBufferStore(e.ctx.Txn(), kv.TempTxnMemBufCap)
+	}
+	if len(e.OnDuplicate) > 0 && !e.IgnoreErr {
+		var err error
+		rows, err = e.batchUpdateDupRows(rows, e.OnDuplicate)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
```

Comment:
We can embed insert logic to `batchUpdateDupRows`, then set last insert ID and finish and return to make the logic clearer.